src/Cli/dotnet/Commands/Test/Terminal/TerminalTestReporter.cs

```diff
@@ -152,28 +154,52 @@ public TerminalTestReporter(IConsole console, TerminalTestReporterOptions option
         _terminalWithProgress = terminalWithProgress;
     }
 
-    public void TestExecutionStarted(DateTimeOffset testStartTime, int workerCount, bool isDiscovery, bool isHelp)
+    public void TestExecutionStarted(DateTimeOffset testStartTime, int workerCount, bool isDiscovery, bool isHelp, bool isRetry)
     {
         _isDiscovery = isDiscovery;
         _isHelp = isHelp;
+        _isRetry = isRetry;
         _testExecutionStartTime = testStartTime;
         _terminalWithProgress.StartShowingProgress(workerCount);
     }
 
-    public void AssemblyRunStarted(string assembly, string? targetFramework, string? architecture, string? executionId)
+    public void AssemblyRunStarted(string assembly, string? targetFramework, string? architecture, string? executionId, string? instanceId)
     {
+        var assemblyRun = GetOrAddAssemblyRun(assembly, targetFramework, architecture, executionId);
+        assemblyRun.Tries.Add(instanceId);
+
+        if (_isRetry)
+        {
+            // When we are retrying the new assembly run should ignore all previously failed tests and
+            // clear all errors. We restarted the run and will retry all failed tests.
+            //
+            // In case of folded dynamic data tests we do not know how many tests we will get in each run,
+            // if more or less, or the same amount as before,and we also will rerun tests that passed previously
+            // because we are unable to run just a single test from that dynamic data source.
+            // This will cause the total number of tests to differ between runs, and there is nothing we can do about it.
+            assemblyRun.TotalTests -= assemblyRun.FailedTests;
+            assemblyRun.RetriedFailedTests += assemblyRun.FailedTests;
+            assemblyRun.FailedTests = 0;
+            assemblyRun.ClearAllMessages();
+        }
```

Comment:
this is the new approach to counting that is more resilient to folded dynamic tests. We don't rely anymore on each tests resetting the previously run tests, but instead  on start of an assembly retry, we make all previously failed tests irrelevant for the current run, and will show them immediately in the rX part in progress indicator.

This way if the second run produces less tests, or different tests, we are not stuck with failed tests from the previous run.